File: src/main/java/com/yuukaze/i18next/TranslatorToolWindowFactory.java
package com.yuukaze.i18next;

import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.wm.ToolWindow;
import com.intellij.openapi.wm.ToolWindowFactory;
import com.intellij.ui.content.Content;
import com.intellij.ui.content.ContentFactory;
import com.yuukaze.i18next.model.SettingsState;
import com.yuukaze.i18next.service.DataStore;
import com.yuukaze.i18next.service.SettingsService;
import com.yuukaze.i18next.service.WindowManager;
import com.yuukaze.i18next.ui.action.*;
import com.yuukaze.i18next.ui.tabs.TableView;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.ResourceBundle;

/**
 * Tool window factory which will represent the entire ui for this plugin.
 *
 * @author marhali
 */
public class TranslatorToolWindowFactory implements ToolWindowFactory {

    @Override
    public void createToolWindowContent(@NotNull Project project, @NotNull ToolWindow toolWindow) {
        ContentFactory contentFactory = ContentFactory.SERVICE.getInstance();

        // Translations table view
        TableView tableView = new TableView(project);
        Content tableContent = contentFactory.createContent(tableView.getRootPanel(),
                ResourceBundle.getBundle("messages").getString("view.table.title"), false);

        toolWindow.getContentManager().addContent(tableContent);

        SettingsState settings = SettingsService.getInstance(project).getState();
        String spreadsheetId = settings.getSpreadSheetId();
        boolean hasSpreadsheetId = spreadsheetId != null && spreadsheetId.length() > 0;

        // ToolWindow Actions (Can be used for every view)
        List<AnAction> actions = new ArrayList<>();
        if (hasSpreadsheetId) {
            actions.add(new SpreadsheetUploadAction());
            actions.add(new SpreadsheetUpdateAction());
        }
        actions.add(new AddAction());
        actions.add(new ReloadAction());
        actions.add(new SettingsAction());
        actions.add(new SearchAction((searchString) -> DataStore.getInstance(project).searchByKey(searchString)));
        toolWindow.setTitleActions(actions);

        // Initialize Window Manager
        WindowManager.getInstance().initialize(toolWindow, tableView);

        // Initialize data store and load from disk
        DataStore store = DataStore.getInstance(project);
        store.addSynchronizer(tableView);
        store.addSynchronizer(tableView.getRootKeyTree());

        store.reloadFromDisk();
    }
}

File: src/main/kotlin/com/yuukaze/i18next/TranslatorToolWindowFactory.kt
package com.yuukaze.i18next

import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.project.Project
import com.intellij.openapi.wm.ToolWindow
import com.intellij.openapi.wm.ToolWindowFactory
import com.intellij.ui.content.ContentFactory
import com.yuukaze.i18next.service.WindowManager
import com.yuukaze.i18next.service.getEasyI18nService
import com.yuukaze.i18next.ui.I18nToolWindow
import com.yuukaze.i18next.ui.Icons
import com.yuukaze.i18next.ui.action.AddAction
import com.yuukaze.i18next.ui.action.SettingsAction
import com.yuukaze.i18next.ui.action.SpreadsheetUpdateAction
import com.yuukaze.i18next.ui.action.SpreadsheetUploadAction
import kotlinx.coroutines.ExperimentalCoroutinesApi
import java.util.*

/**
 * Tool window factory which will represent the entire ui for this plugin.
 */
class TranslatorToolWindowFactory : ToolWindowFactory {
  override fun init(toolWindow: ToolWindow) {
    toolWindow.setIcon(Icons.ToolWindowIcon)
  }

  internal companion object {
    val mapper = mutableMapOf<Project, I18nToolWindow>()
    fun getToolWindow(project: Project): I18nToolWindow = mapper[project]!!
  }

  @ExperimentalCoroutinesApi
  override fun createToolWindowContent(
    project: Project,
    toolWindow: ToolWindow
  ) {
    // Translations table view
    val contentFactory = ContentFactory.SERVICE.getInstance()
    val toolWindowComponent = I18nToolWindow(project, toolWindow)
    //add to map
    mapper[project] = toolWindowComponent

    val tableContent = contentFactory.createContent(
      toolWindowComponent.rootPanel,
      ResourceBundle.getBundle("messages").getString("view.table.title"), false
    )
    toolWindow.contentManager.addContent(tableContent)
    val settings = project.getEasyI18nService().state
    val spreadsheetId = settings.spreadSheetId
    val hasSpreadsheetId = spreadsheetId.isNotEmpty()

    // ToolWindow Actions (Can be used for every view)
    val actions: MutableList<AnAction> = ArrayList()
    if (hasSpreadsheetId) {
      actions.add(SpreadsheetUploadAction())
      actions.add(SpreadsheetUpdateAction())
    }
    actions.add(AddAction())
    actions.add(SettingsAction())
    toolWindow.setTitleActions(actions)

    // Initialize Window Manager
    WindowManager.getInstance()
      .initialize(toolWindow, toolWindowComponent.tableView)

    // Initialize data store and load from disk
    val store = project.getEasyI18nService().dataStore
    store.reloadFromDisk()
  }
}

@ExperimentalCoroutinesApi
fun Project.openToolWindow() =
  TranslatorToolWindowFactory.getToolWindow(this).apply { toolWindow.show() }